kcidb_io/schema/v1.py

```diff
@@ -455,6 +455,410 @@
+                },
+                "description": {
+                    "type": "string",
+                    "description":
+                        "Human-readable description of the revision. "
+                        "E.g. a release version, or the subject of a patchset message."
+                },
+                "publishing_time": {
+                    "type": "string",
+                    "format": "date-time",
+                    "description":
+                        "The time the revision was made public. E.g. the timestamp "
+                        "on a patch message, a commit, or a tag.",
+                },
+                "discovery_time": {
+                    "type": "string",
+                    "format": "date-time",
+                    "description":
+                        "The time the revision was discovered by the CI system. "
+                        "E.g. the time the CI system found a patch message, or "
+                        "noticed a new commit or a new tag in a git repo.",
+                },
+                "contacts": {
+                    "type": "array",
+                    "description":
+                        "List of e-mail addresses of contacts concerned with "
+                        "this revision, such as authors, reviewers, and mail lists",
+                    "items": {
+                        "type": "string",
+                        "description":
+                            "An e-mail address of a contact concerned with this "
+                            "revision, e.g. an author, a reviewer, or a mail list, "
+                            "as in https://tools.ietf.org/html/rfc5322#section-3.4"
+                    },
+                },
+                "log_url": {
+                    "type": "string",
+                    "format": "uri",
+                    "description":
+                        "The URL of the log file of the attempt to construct this "
+                        "revision from its parts. E.g. 'git am' output.",
+                },
+                "valid": {
+                    "type": "boolean",
+                    "description":
+                        "True if the revision is valid, i.e. if its parts could be "
+                        "combined. False if not, e.g. if its patches failed to apply."
+                },
+                "misc": {
+                    "type": "object",
+                    "description":
+                        "Miscellaneous extra data about the revision",
+                },
+            },
+            "additionalProperties": False,
+
+            "description":
+                "A revision of the tested code.\n"
+                "\n"
+                "Represents a way the tested source code could be obtained. E.g. "
+                "checking out a particular commit from a git repo, and applying a "
+                "set of patches on top.",
+            "$id": "/schemas/revision",
```

Comment:
`$id` is used to define cannonical URI with pointer fragment. [check it out](https://json-schema.org/draft/2019-09/json-schema-core.html#idExamples)
Perusing [this document](https://json-schema.org/understanding-json-schema/structuring.html#id20) gave me the idea for using `$id`

Reply:
Re-reading the doc shows that there was no mention of using `$id` in `$def`. I just inferred. I'll make the changes.

Reply:
Good job figuring this all out, and I understand this could be useful if we want to keep schemas in separate files and then bundle them automatically. However, we do **not** want to put the schema in separate files so far (i.e. we don't want to "unbundle" it). It's easier to deal with it in a single file thus far.

Using this mechanism would be clever, but too involved for our needs right now. It is not obvious how this works, and reading and understanding the article you linked takes effort and time. So let's not subject others and future us to this unnecessarily, and use the relatively-obvious fragment URIs pointing to the schemas using "JSON Pointers". I.e. just put schemas under `"$defs"`, and use `"$ref": "#/$defs/resource"` and similar.

We control the entirety of the schema, and we would be able to change the references and IDs if we ever need it.
Thank you.

Reply:
The article does take a lot of effort and time. I'll submit a commit to that effect.

Also, what should I do about the failing test?

Reply:
> Also, what should I do about the failing test?

Wrap the lines. Rewrap the documentation paragraphs as necessary. Are there any particular lines that are giving you trouble?

If this turns out too difficult or ugly, we could still just leave the schema pieces in the variables as they are, and paste *them* into `$defs` instead.